[17.0][MIG] contract_sale_generation: Migration to 17.0

javierjcf opened the pull request:
This PR migrates the module `contract_sale_generation` from 16.0 to 17.0.
Superseed #1087.

angelmoya committed:
ADD contract_sale_generation module

moylop260 committed:
[FIX] contract_payment_auto: Fix freze wkhtmltopdf freezed

angelmoya committed:
[FIX] contract_sale_generation: Isolate sale generation (#118)

Closes #117

pedrobaeza committed:
[FIX] contract_sale_generation: Useless test

pedrobaeza committed:
[10.0][FIX+IMP] contract: Improve usability and don't fail on wrong data (#130)

* [FIX+IMP] contract: Improve usability and don't fail on wrong data

* Cron create invoices masked for avoiding silent errors
* New constraints for assuring data consistency
* UI helps for entering consistent data
* Spanish translation
* Remove double company_id field on form

* [FIX] contract_sale_generation: Adapt tests to upstream contract

flotho committed:
[FIX] contract_sale_generation: it doesn't create sales (#141)

* Change the method called in the view
* Complete the create_invoice method
* Bump version + authoring
* Correct bad call of method
  Small Documentation
* Add super call in python test
* FIX bad field names causing bad quantities in sale.order.line

NL66278 committed:
[FIX] contract. Optimize insertion of dates in invoice and lines.

mvrodriguez committed:
Translated using Weblate (Galician)

Currently translated at 100.0% (10 of 10 strings)

Translation: contract-10.0/contract-10.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-10-0/contract-10-0-contract_sale_generation/gl/

marylla committed:
Translated using Weblate (German)

Currently translated at 100.0% (10 of 10 strings)

Translation: contract-10.0/contract-10.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-10-0/contract-10-0-contract_sale_generation/de/

jarmokortetjarvi committed:
Translated using Weblate (Finnish)

Currently translated at 100.0% (10 of 10 strings)

Translation: contract-10.0/contract-10.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-10-0/contract-10-0-contract_sale_generation/fi/

murtuzasaleh committed:
[WIP][12.0][MIG] contract_sale_generation

britoederr committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-12.0/contract-12.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-contract_sale_generation/pt_BR/

flotho committed:
add display type in default values

remi-filament committed:
Translated using Weblate (French)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-12.0/contract-12.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-contract_sale_generation/fr/

mkxmawilix committed:
[12.0][FIX] contract: Incorrect product UOM qty dict key for the sale.order.line

The product UOM qty was not correctly set on the Sale Order Line
During the creation of a recurring Order from a Contract, the given quantity on the Contract Line was not correctly set on the created Sale Order Line.

Freni-OSI committed:
[14.0][FIX] Travis

Freni-OSI committed:
[14.0][MIG] contract_sale_generation

rousseldenis committed:
[14.0][IMP] contract_sale_generation

Use the generation_type field defined now in contract base module.

Improve tests

rousseldenis committed:
[IMP] contract_sale_generation: Improve view

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Transmit analytic account

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Add hook method to retrieve sale line values

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Align method on Odoo conventions

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Take into account new method for cron

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Remove deprecated domain onchange

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Split test and common

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Remove unused code

ibuioli committed:
Added translation using Weblate (Spanish (Argentina))

rousseldenis committed:
[14.0][FIX] contract_sale_generation: Recompute next recurring date after generation on contract level

As in contract module, recompute recurring_next_date after sale generation on contract level

rousseldenis committed:
[14.0][FIX] contract_sale_generation: Add start date on contract level (for recurrency on contract)

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Add tests with another recurrency

Noel000 committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-14.0/contract-14.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-14-0/contract-14-0-contract_sale_generation/ca/

scigghia committed:
Translated using Weblate (Italian)

Currently translated at 73.3% (11 of 15 strings)

Translation: contract-14.0/contract-14.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-14-0/contract-14-0-contract_sale_generation/it/

ntsirintanis committed:
[MIG] contract_sale_generation: Migration to 15.0

victoralmau committed:
[FIX] contract_sale_generation: Change SavepointCase to TransactionCase

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 93.3% (14 of 15 strings)

Translation: contract-15.0/contract-15.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-15-0/contract-15-0-contract_sale_generation/it/

JesusValdez96 committed:
[MIG] contract_sale_generation: Migration to 16.0

Commit:
[UPD] Update contract_sale_generation.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 83.3% (10 of 12 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/pt_BR/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (12 of 12 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/es/

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (12 of 12 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/pt_BR/

aliciagaarzo committed:
[16.0][IMP] contract and contract_sale_generation: Add translations

Commit:
[UPD] Update contract_sale_generation.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/es/

ThijsvOers committed:
Translated using Weblate (Dutch)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/nl/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/it/

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/pt_BR/

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 86.6% (13 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/hr/

luisDIXMIT committed:
[FIX] contract_sale_generation: Post-install test + fallback to load CoA

OCA-git-bot committed:
[BOT] post-merge updates

javierjcf committed:
[IMP] contract_sale_generation: pre-commit auto fixes

javierjcf committed:
[MIG] contract_sale_generation: Migration to 17.0

javierjcf commented:
Hi, 

I dont understand new codecov errors. It seems like is detecting lines of code not tested, but in my migration commit i only edit xml files. Is a simple migration... why is failing coverage if i did not add any python line.

Is mandatory to get this coverage in order to merge this PR?

rousseldenis commented:
> I dont understand new codecov errors. It seems like is detecting lines of code not tested, but in my migration commit i only edit xml files. Is a simple migration... why is failing coverage if i did not add any python line.

@javierjcf As for now (I'm not fond of :-) ), we start from blank base new branches. So, the module you migrate is not present. Coverage will do the full change tests. If you look at your PR, it adds the whole module and not only migration commit.

rousseldenis commented:
/ocabot migration contract_sale_generation

rousseldenis commented:
> > I dont understand new codecov errors. It seems like is detecting lines of code not tested, but in my migration commit i only edit xml files. Is a simple migration... why is failing coverage if i did not add any python line.
> 
> @javierjcf As for now (I'm not fond of :-) ), we start from blank base new branches. So, the module you migrate is not present. Coverage will do the full change tests. If you look at your PR, it adds the whole module and not only migration commit.

@javierjcf It seems tests are not running at all. Could you check locally (it can maybe be the accounting modules that causes problem)

javierjcf commented:
Thanks @rousseldenis 

Yes,  tests weren't working, there are references in the test cases that are no longer valid. I will try to fix them throughout the week.

javierjcf commented:
Hi, I need help with a test error. 

`AssertionError: can't write on invisible field 'quantity'
`

It doesn't occur in my local environment with a fresh database. However, I did face the same error when accessing contract_line_ids. That field was indeed invisible, and I had to explicitly set contract_form.line_recurrence = True. With that, the tests passed.

In CI, the error is triggered by the quantity field. These fields don't have any invisibility modifiers, except for the parent group, which depends on display_type. But if I add display_type to the test to make quantity writable, I get an error saying that display_type is invisible (which is true).

Could another module be interfering? In my local environment I only have contract_sale_generation and its dependencies installed, not the entire OCA...

![image](https://github.com/user-attachments/assets/88654c40-a048-4604-abff-66f028c705f5)


I see that if I install the rest of the modules from this repository, I also get the same error locally.
Is it expected that if another module adds an invisibility modifier to a field, or to the part of the view where a field is located, it can cause the tests to fail?

Am I supposed to always run the tests with the other modules in the repository? Or with the whole OCA?

javierjcf commented:
I found that module **contract_variable_qty** is doing:
``` 
<field name="quantity" position="attributes">
    <attribute name="invisible">"qty_type != 'fixed'"</attribute>
</field>
```

I don't think the solution is to worry about fields introduced by other modules, especially when the default is correctly set...
Even if I'm explicit by doing:

`line_form.qty_type = 'fixed'`

It still fails the same way.

How can I make the quantity field always visible from the test?

javierjcf commented:
I was trying to understand why the field is invisible in the test, but it doesn’t seem to be.

![image](https://github.com/user-attachments/assets/6321dfc9-9686-4056-872f-b0427722b057)

The invisibility condition is set to False, yet the error is still being raised. How is this possible?

I was debugging and i see this:
![image](https://github.com/user-attachments/assets/cf6cd042-73ca-4224-b5a2-e25d5b0e0a5f)

The expresion `'(display_type) or ("qty_type != \'fixed\'")'` seems wrong. Its fault of this:
```
       <xpath expr="//field[@name='quantity']" position="attributes">
                <attribute name="required">"qty_type == 'fixed'"</attribute>
                <attribute name="invisible">"qty_type != 'fixed'"</attribute>
            </xpath>
```

Double quote makes not sense in the <attribute> tag.

I will make a PR to fix that bug.

javierjcf commented:
I tested in local with #1242 and test passed. To pass the test here we need #1243 i think

ofonseca-pyming reviewed (approved):
LGTM

rousseldenis commented on contract_sale_generation/tests/test_contract_sale.py:
```diff
@@ -4,12 +4,11 @@
 
 
-class TestContractSale(ContractSaleCommon, TransactionCase):
+class TestContractSale(ContractSaleCommon):
```
@javierjcf Don't do these changes. They are not needed. Thanks

Wodran14 commented:
Any updates on this module?

adasatorres reviewed (approved):
LGTM, Code review.

rrebollo commented:
@javierjcf runboat is not accesible, you could do a rebase. @rousseldenis what we need to get this done?

rousseldenis commented:
@javierjcf Could you check tests?

javierjcf commented:
> @javierjcf Could you check tests?

"I tried, but I don’t understand what’s failing (it was working before; only the runbiat part was wrong).

Analytic accounting has been tricky for me since the v16 changes. In any case, the prepare_sale function propagates the analytic account to the orders (which is now called group_id in the contract). I noticed it’s possible it isn't entering this method because it lacks a date_ref. I’ve updated the test to call the private version of the method, _recurring_create_sale, so I can pass the date_ref and ensure prepare_sale isn't skipped.

Even so, it still throws the same error.

In the v18 PR, I removed the line that checks for that—not a great solution. However, in a private project using the v18 module, the account is propagating correctly. I’m missing something regarding how the test is designed, but I don't have time this week to look into it in depth. If anyone knows why this is happening, I can change it immediately.

Otherwise, I’ll try to review it when I have more time available."

rrebollo commented:
@javierjcf Please avoid using the "remove conflicting call" approach. I recommend reverting it in the v18 PR as well. I'm currently working on a migration that requires this addition. I'll investigate the failing test and share my findings here once I have them.

javierjcf commented:
@rrebollo I tried to revert, but in v18 is a key errror. Not analytic_account_id on sale.order

<img width="1393" height="213" alt="image" src="https://github.com/user-attachments/assets/726d09f1-1b74-453a-a5f8-66551c244c32" />

I'm not sure about analytic plan/distribution and the right way to check it. I think test shoud new check analitic_distribution json in lines instead of sale.order level?? i'm not sure...

In v17 we still have a general analytic_account_id on sale.order, but makes not sense to me. In v18 this field is removed.

If we found a solution heare in v17 i will fix 18 PR as well

javierjcf commented:
@rrebollo 

I think I found the bug. The analytic account is being set at the contract header level in the common.py file. However, since the lines are created using Form() and group_id is a computed field, the value is lost. Furthermore, the lines in v17 do not have the analytic distribution in the view; perhaps it should be added...

Anyway, since I know the group_id field can be lost, I am re-setting it after using Form() to create the lines. (I don't understand why the test data setup mixes direct value assignment in cls with line creation using Form()).

This way, cls.contract preserves the analytic account that was initially established:

https://github.com/OCA/contract/blob/d1475c036bb2e787b9306348f3b702eb30870222/contract_sale_generation/tests/common.py#L134

I don't use v17 or the analytic module myself, so I'm not entirely sure if this is the correct way to test it. It seems more like the new analytic_distribution widget should be at the line level (and visible too, otherwise we can't use Form() to assign a value to it) and then verify that the contract lines have the same analytic distribution as the generated sales order lines.

In v18, I think it will have to be this way since the analytic account field doesn't even exist at the sale.order level anymore.

rrebollo commented:
@javierjcf To validate your approach—both for the tests and the overall migration—I recommend looking at merged addons in v17.0 branches of OCA repositories related to analytic workflow. A focused code search should be sufficient.

You can then compare their implementation with your approach. Thank you for all the effort you're putting into this. I haven't had a chance to review it today.

javierjcf commented:
Hi @rrebollo, the functionality intended for this test (propagating the analytic account from the contract to the sale) is already being tested correctly. Regarding the analytic distribution at the contract line level, that logic is in the 'contract' module, which has already been merged. The base module handles this part correctly, although it doesn't seem to be tested at the line level.

Functionally, this module seems correct; the only issue was the test. While the test could be improved, I think it's fine as it is for the migration PR.

rrebollo commented on contract_sale_generation/models/abstract_contract.py:
```diff
@@ -0,0 +1,15 @@
+    sale_autoconfirm = fields.Boolean()
+
+    generation_type = fields.Selection(
+        selection_add=[("sale", "Sale")],
```
```suggestion
        selection_add=[("sale", "Sale")], ondelete={"sales": "cascade"}
```

rrebollo commented on contract_sale_generation/models/contract.py:
```diff
@@ -0,0 +1,133 @@
+        sales_values = self._prepare_recurring_sales_values(date_ref)
+        sale_orders = self.env["sale.order"].create(sales_values)
+        sale_orders_to_confirm = sale_orders.filtered(
+            lambda sale: sale.contract_auto_confirm
```
```suggestion
            "contract_auto_confirm"
```
